// unicoen.ts/dist/interpreter/Interpreter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Interpreter = void 0;
class Interpreter {
    constructor(engine, mapper) {
        this.engine = engine;
        this.mapper = mapper;
    }
    setDebugMode(enable) {
        this.engine.setDebugMode(enable);
    }
    getCurrentValue() {
        return this.engine.getCurrentValue();
    }
    getCurrentExpr() {
        return this.engine.getCurrentExpr();
    }
    getCurrentState() {
        return this.engine.getCurrentState();
    }
    getStdout() {
        return this.engine.getStdout();
    }
    stdin(text) {
        this.engine.stdin(text);
    }
    getIsWaitingForStdin() {
        return this.engine.getIsWaitingForStdin();
    }
    isStepExecutionRunning() {
        return this.engine.isStepExecutionRunning();
    }
    setFileList(filelist) {
        this.engine.setFileList(filelist);
    }
    preProcess(code) {
        return code;
    }
    startStepExecution(code) {
        const preProcessedCode = this.preProcess(code);
        const tree = this.mapper.parseToUniTree(preProcessedCode);
        return this.engine.startStepExecution(tree);
    }
    stepExecute() {
        return this.engine.stepExecute();
    }
    execute(code) {
        const preProcessedCode = this.preProcess(code);
        const tree = this.mapper.parseToUniTree(preProcessedCode);
        return this.engine.execute(tree);
    }
    checkSyntaxError(code) {
        return this.mapper.checkSyntaxError(code);
    }
}
exports.Interpreter = Interpreter;
