// unicoen.ts/dist/node/UniBinOp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniBinOp = void 0;
const UniExpr_1 = require("./UniExpr");
class UniBinOp extends UniExpr_1.UniExpr {
    constructor(operator, left, right) {
        super();
        if (operator === undefined && left === undefined && right === undefined) {
            this.operator = null;
            this.left = null;
            this.right = null;
        }
        else if (operator === undefined || left === undefined || right === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.operator = operator;
            this.left = left;
            this.right = right;
        }
        this.fields.set('operator', String);
        this.fields.set('left', UniExpr_1.UniExpr);
        this.fields.set('right', UniExpr_1.UniExpr);
    }
    toString() {
        return 'BinOp(' + this.operator + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniBinOp))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.operator == null ? that.operator == null : this.operator === that.operator)
            && (this.left == null ? that.left == null : this.left.equals(that.left))
            && (this.right == null ? that.right == null : this.right.equals(that.right));
    }
    merge(that) {
        super.merge(that);
        if (that.operator != null) {
            this.operator = that.operator;
        }
        if (that.left != null) {
            this.left = that.left;
        }
        if (that.right != null) {
            this.right = that.right;
        }
    }
}
exports.UniBinOp = UniBinOp;
