// unicoen.ts/dist/node/UniBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniBlock = void 0;
const UniStatement_1 = require("./UniStatement");
class UniBlock extends UniStatement_1.UniStatement {
    constructor(blockLabel, body) {
        super();
        if (blockLabel === undefined && body === undefined) {
            this.blockLabel = null;
            this.body = [];
        }
        else if (blockLabel === undefined || body === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.blockLabel = blockLabel;
            this.body = body;
        }
        this.fields.set('blockLabel', String);
        this.fields.set('body', UniStatement_1.UniStatement);
    }
    toString() {
        return 'Block(' + this.blockLabel + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniBlock))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.blockLabel == null ? that.blockLabel == null : this.blockLabel === that.blockLabel)
            && (this.body == null ? that.body == null : this.body.equals(that.body));
    }
    merge(that) {
        super.merge(that);
        if (that.blockLabel != null) {
            this.blockLabel = that.blockLabel;
        }
        if (that.body != null) {
            if (this.body == null) {
                this.body = that.body;
            }
            else {
                this.body.push(...that.body);
            }
        }
    }
}
exports.UniBlock = UniBlock;
