// unicoen.ts/dist/node/UniImportDec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniImportDec = void 0;
const UniDecralation_1 = require("./UniDecralation");
class UniImportDec extends UniDecralation_1.UniDecralation {
    constructor(name, ondemand) {
        super();
        if (name === undefined && ondemand === undefined) {
            this.name = null;
            this.ondemand = null;
        }
        else if (name === undefined || ondemand === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.name = name;
            this.ondemand = ondemand;
        }
        this.fields.set('name', String);
        this.fields.set('ondemand', String);
    }
    toString() {
        return 'ImportDec(' + this.name + ', ' + this.ondemand + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniImportDec))
            return false;
        const that = obj;
        return (super.equals(that) &&
            (this.name == null ? that.name == null : this.name === that.name) &&
            (this.ondemand == null ? that.ondemand == null : this.ondemand === that.ondemand));
    }
    merge(that) {
        super.merge(that);
        if (that.name != null) {
            this.name = that.name;
        }
        if (that.ondemand != null) {
            this.ondemand = that.ondemand;
        }
    }
}
exports.UniImportDec = UniImportDec;
