// unicoen.ts/dist/node/UniNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniNode = void 0;
const CodeRange_1 = require("../node_helper/CodeRange");
require('../node_helper/Extension');
class UniNode {
    constructor(comments, codeRange) {
        if (comments === undefined && codeRange === undefined) {
            this.comments = [];
            this.codeRange = null;
        }
        else if (comments === undefined || codeRange === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.comments = comments;
            this.codeRange = codeRange;
        }
        this.fields = new Map();
        this.fields.set('comments', String);
        this.fields.set('codeRange', CodeRange_1.CodeRange);
    }
    toString() {
        return 'Node(' + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniNode))
            return false;
        const that = obj;
        return that !== null;
    }
    merge(that) {
        if (that.comments != null) {
            if (this.comments == null) {
                this.comments = that.comments;
            }
            else {
                this.comments.push(...that.comments);
            }
        }
        if (that.codeRange != null) {
            this.codeRange = that.codeRange;
        }
    }
}
exports.UniNode = UniNode;
