// unicoen.ts/dist/node/UniNew.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniNew = void 0;
const UniExpr_1 = require("./UniExpr");
class UniNew extends UniExpr_1.UniExpr {
    constructor(type, args) {
        super();
        if (type === undefined && args === undefined) {
            this.type = null;
            this.args = [];
        }
        else if (type === undefined || args === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.type = type;
            this.args = args;
        }
        this.fields.set('type', String);
        this.fields.set('args', UniExpr_1.UniExpr);
    }
    toString() {
        return 'New(' + this.type + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniNew))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.type == null ? that.type == null : this.type === that.type)
            && (this.args == null ? that.args == null : this.args.equals(that.args));
    }
    merge(that) {
        super.merge(that);
        if (that.type != null) {
            this.type = that.type;
        }
        if (that.args != null) {
            if (this.args == null) {
                this.args = that.args;
            }
            else {
                this.args.push(...that.args);
            }
        }
    }
}
exports.UniNew = UniNew;
