// unicoen.ts/dist/node/UniSwitch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniSwitch = void 0;
const UniStatement_1 = require("./UniStatement");
const UniExpr_1 = require("./UniExpr");
const UniSwitchUnit_1 = require("./UniSwitchUnit");
class UniSwitch extends UniStatement_1.UniStatement {
    constructor(cond, cases) {
        super();
        if (cond === undefined && cases === undefined) {
            this.cond = null;
            this.cases = [];
        }
        else if (cond === undefined || cases === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.cond = cond;
            this.cases = cases;
        }
        this.fields.set('cond', UniExpr_1.UniExpr);
        this.fields.set('cases', UniSwitchUnit_1.UniSwitchUnit);
    }
    toString() {
        return 'Switch(' + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniSwitch))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.cond == null ? that.cond == null : this.cond.equals(that.cond))
            && (this.cases == null ? that.cases == null : this.cases.equals(that.cases));
    }
    merge(that) {
        super.merge(that);
        if (that.cond != null) {
            this.cond = that.cond;
        }
        if (that.cases != null) {
            if (this.cases == null) {
                this.cases = that.cases;
            }
            else {
                this.cases.push(...that.cases);
            }
        }
    }
}
exports.UniSwitch = UniSwitch;
