// unicoen.ts/dist/node/UniUnaryOp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniUnaryOp = void 0;
const UniExpr_1 = require("./UniExpr");
class UniUnaryOp extends UniExpr_1.UniExpr {
    constructor(operator, expr) {
        super();
        if (operator === undefined && expr === undefined) {
            this.operator = null;
            this.expr = null;
        }
        else if (operator === undefined || expr === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.operator = operator;
            this.expr = expr;
        }
        this.fields.set('operator', String);
        this.fields.set('expr', UniExpr_1.UniExpr);
    }
    toString() {
        return 'UnaryOp(' + this.operator + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniUnaryOp))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.operator == null ? that.operator == null : this.operator === that.operator)
            && (this.expr == null ? that.expr == null : this.expr.equals(that.expr));
    }
    merge(that) {
        super.merge(that);
        if (that.operator != null) {
            this.operator = that.operator;
        }
        if (that.expr != null) {
            this.expr = that.expr;
        }
    }
}
exports.UniUnaryOp = UniUnaryOp;
