// unicoen.ts/dist/node/UniVariableDec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniVariableDec = void 0;
const UniDecralation_1 = require("./UniDecralation");
const UniVariableDef_1 = require("./UniVariableDef");
class UniVariableDec extends UniDecralation_1.UniDecralation {
    constructor(modifiers, type, variables) {
        super();
        if (modifiers === undefined && type === undefined && variables === undefined) {
            this.modifiers = [];
            this.type = null;
            this.variables = [];
        }
        else if (modifiers === undefined || type === undefined || variables === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.modifiers = modifiers;
            this.type = type;
            this.variables = variables;
        }
        this.fields.set('modifiers', String);
        this.fields.set('type', String);
        this.fields.set('variables', UniVariableDef_1.UniVariableDef);
    }
    toString() {
        return 'VariableDec(' + this.type + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniVariableDec))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.modifiers == null ? that.modifiers == null : this.modifiers.equals(that.modifiers))
            && (this.type == null ? that.type == null : this.type === that.type)
            && (this.variables == null ? that.variables == null : this.variables.equals(that.variables));
    }
    merge(that) {
        super.merge(that);
        if (that.modifiers != null) {
            if (this.modifiers == null) {
                this.modifiers = that.modifiers;
            }
            else {
                this.modifiers.push(...that.modifiers);
            }
        }
        if (that.type != null) {
            this.type = that.type;
        }
        if (that.variables != null) {
            if (this.variables == null) {
                this.variables = that.variables;
            }
            else {
                this.variables.push(...that.variables);
            }
        }
    }
}
exports.UniVariableDec = UniVariableDec;
