// unicoen.ts/dist/node/UniWhile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniWhile = void 0;
const UniStatement_1 = require("./UniStatement");
const UniExpr_1 = require("./UniExpr");
class UniWhile extends UniStatement_1.UniStatement {
    constructor(cond, statement) {
        super();
        if (cond === undefined && statement === undefined) {
            this.cond = null;
            this.statement = null;
        }
        else if (cond === undefined || statement === undefined) {
            throw new Error('invalid arguments');
        }
        else {
            this.cond = cond;
            this.statement = statement;
        }
        this.fields.set('cond', UniExpr_1.UniExpr);
        this.fields.set('statement', UniStatement_1.UniStatement);
    }
    toString() {
        return 'While(' + ')';
    }
    equals(obj) {
        if (obj == null || !(obj instanceof UniWhile))
            return false;
        const that = obj;
        return super.equals(that)
            && (this.cond == null ? that.cond == null : this.cond.equals(that.cond))
            && (this.statement == null ? that.statement == null : this.statement.equals(that.statement));
    }
    merge(that) {
        super.merge(that);
        if (that.cond != null) {
            this.cond = that.cond;
        }
        if (that.statement != null) {
            this.statement = that.statement;
        }
    }
}
exports.UniWhile = UniWhile;
